metabase/metabase
# Fix float-to-int type coercion for CSV appends using declarative maps

### Description

This follows on from the https://github.com/metabase/metabase/pull/39493 work, to fix a regression on float-to-int coercion that can cause needless column alteration. It also subsumes a simplified version of the [hiccup DSL](https://github.com/metabase/metabase/pull/39793), by removing the coupling to the ordering of child-to-parent edges.

This approach is an alternative to the [lattice approach](https://github.com/metabase/metabase/pull/39724), and instead takes a more domain centered and explicit approach. 

Removing `::explicit-int` is playing with fire in some ways, introducing for example the gotcha that `1.0` is a member of `::int` (in a column-type setting), but not a member of `::int` (in a value-type setting). It causes our hierarchy in the code to reflect a sub-typing relationship only for column-types, and the implicit, divergent hierarchy for value-types is never visualized. 

While I like the explicit nature of this interface, it is less general and extensible, with subtle gotchas. Overall I think it could work well for the foreseeable future, and suspect there is a way to heal the model if we stop overloading the keywords for both types. 

### How to verify

See [this test case.](https://github.com/metabase/metabase/pull/40275/files#diff-d72e1e62dbb726e6e563e8b2caec58e6cac2bbb9a650f10f0f42ecf2f2d00debL1793-R1794)
